### PR TITLE
[v1.13] node: Fix IP removal from ipset on node updates

### PR DESCRIPTION
**Pull request by qmonnet**
In CI, on upgrade/downgrade tests for IPsec, we have observed regular failures on some workflows for the node-to-node-encryption test. In this test, we look for some unencrypted packets from a client pod to a host on a distinct node; but sometimes, these packets do not show up.

Investigation showed that the packets do show up, but not with the expected IP address, thus failing to be caught by the tcpdump filter in use. In the failure state, instead of having the client pod IP as source address, packets have the origin host IP as source address. This is not supposed to happen: we use an iptables rule to exclude packets from masquerading in such case. Why then masquerading still occur? This is because the rule doesn't apply, the relevant host IP being missing in the ipset used by the rule.

Looking at the agent, the relevant node IP is missing from the ipset because on node updates, we sometimes remove it, when we should not. And the reason for that is that we're not comparing the right values to decide whether this removal should happen. Indeed, in the following:

    !slices.Contains(ipsAdded, address.IP.String())

The `address.IP.String()` represents a mere IP address, when the values added to ipsAdded are `netip.Prefix.String()`, containing the IP followed by a CIDR mask with the slash notation. For an example, `ipsAdded` can be `[172.19.0.2/32 fc00:c111::2/128]`, and `address.IP.String()` can be `172.19.0.2` (no CIDR mask): the search fails and we do remove the IP from the ipset, even if we just added it above in the function.

This commit fixes the comparison to ensure that we're looking for the relevant values in the slice, which in turns fixes the set of IPs contained in the ipset, and applies the masquerading exclusion to the relevant packets.

Related: https://github.com/cilium/cilium/issues/29351

### CODE REVIEW

**Comment by qmonnet** [comment]
/test-backport-1.13

**Comment by qmonnet** [comment]
/test-backport-1.13

**Comment by brb** [comment]
Very nice find! I think `manager_test.go` deserves some new test cases :-)

**Comment by qmonnet** [comment]
Ah right re-enabling `node-to-node-encryption` is in the GH Action, so I need a dedicated commit to test the change

**Comment by qmonnet** [comment]
> Very nice find! I think `manager_test.go` deserves some new test cases :-)

@brb Please take a look at https://github.com/cilium/cilium/pull/29898/commits/2dfda5e47b5c13b4ab05a2c75abce35cdf4a14db

Note: apparently it's too early to re-enable `node-to-node-encryption`, because the issue seems present on 1.12 as well and we're hitting it on downgrades. I'll try to port this to 1.12 as well, and once we have it merged there we can re-enable `node-to-node-encryption` on this branch. Meanwhile, we can probably take a look at merging the first two commits, to make downgrade work from the 1.14 branch.

**Comment by qmonnet** [comment]
/test-backport-1.13

**Comment by qmonnet** [comment]
> because the issue seems present on 1.12 as well

Hmm in fact it shouldn't, the issue was seemingly introduced by 454df0c065bd94aa30a9d703732d64a24d3b6073 which was added in 1.13 :thinking:

**Comment by qmonnet** [comment]
/test-backport-1.13

**Comment by qmonnet** [comment]
/test-backport-1.13

**Comment by qmonnet** [comment]
Travis fails because the image running there doesn't have the `ipset` binary :disappointed:

**Comment by brb** [comment]
> Travis fails because the image running there doesn't have the ipset binary 😞

You should be able to install it (check `.travis/prepare.sh`).

**Review by brb — APPROVED** [review]
Nice! :rocket:

**Review by gandro — APPROVED** [review]
The pure nodemanager change looks good to me. Do we need to forward-port this to `main`?

**Comment by qmonnet** [comment]
> Do we need to forward-port this to `main`?

@gandro No, the bug was introduced in 454df0c065bd94aa30a9d703732d64a24d3b6073, but it seems you fixed it by reworking this code in f2a43127882db10017d2a80670352f518e50e7a7 (which is why I asked the review from you).

I intend to port the test, though.

**Comment by qmonnet** [comment]
> Note: apparently it's too early to re-enable `node-to-node-encryption`, because the issue seems present on 1.12 as well and we're hitting it on downgrades.

This was wrong.

- 1.12 is not affected, the commit that introduced the issue (f2a43127882db10017d2a80670352f518e50e7a7) was introduced in 1.13. And the workflow didin't, in fact, fail with 1.12.
- The failure observed was for the downgrade to the latest _patch_ release, 1.13.z, which does contain the issue and that won't be fixed until 1) we merge the current PR and 2) we cut a new minor release after that.

So we can merge the fix on 1.13, but we won't be able to re-enable `node-to-node-encryption`, either on 1.13 (we'd fail on downgrades to the latest patch release) or in 1.14 (we'd fail on downgrade to the latest minor version), until we've cut a new minor release on 1.13 containing the fix. (Luckily, the release manager for January is someone I know well.)

So I think we're ready to merge the fix itself. I'm trying to make the runtime test pass as well, if it doesn't I'll take it out for a follow-up PR so we can get the fix in.

**Comment by qmonnet** [comment]
I needed to set the test as privileged so that we can run `ipset` successfully, let's give it another try...

**Comment by qmonnet** [comment]
/test-backport-1.13

**Comment by qmonnet** [comment]
/test-backport-1.13

**Comment by qmonnet** [comment]
/test-1.23-4.19

**Comment by qmonnet** [comment]
Travis is happy this time

**Comment by julianwiedmann** [comment]
Just to leave some GH bread-crumbs - we're saying this was introduced with https://github.com/cilium/cilium/pull/19765, and fixed by https://github.com/cilium/cilium/pull/23208.

**Review by julianwiedmann — APPROVED** [review]

**Pull request merged**
